src/javaharness/java/arcs/crdt/CRDTCollection.java

```diff
@@ -0,0 +1,174 @@
+    }
+  }
+
+  @Override
+  public CollectionData<T> getData() { return model; }
+
+  @Override
+  public CRDTConsumerType getParticleView() {
+    return new RawCollection(
+        model.values.values().stream().map(v -> v.value).collect(Collectors.toList()));
+  }
+
+  private boolean add(T value, String key, VersionMap version) {
+    // Only accept an add if it is immediately consecutive to the clock for that actor.
+    int expectedClockValue = model.version.getOrDefault(key, 0).intValue() + 1;
+    if (expectedClockValue != version.getOrDefault(key, 0)) {
+      return false;
+    }
+    this.model.version.put(key, version.getOrDefault(key, 0));
+    VersionMap previousVersion = model.values.containsKey(value.getId())
+        ? model.values.get(value.getId()).version : new VersionMap();
+    model.values.put(value.getId(), new VersionedValue(value, mergeVersions(version, previousVersion)));
+    return true;
+  }
+
+  private boolean remove(T value, String key, VersionMap version) {
+    if (!this.model.values.containsKey(value.getId())) {
+      return false;
+    }
+    int clockValue = version.getOrDefault(key, 0).intValue();
+    // Removes do not increment the clock.
+    int expectedClockValue = model.version.getOrDefault(key, 0).intValue();
+    if (expectedClockValue != clockValue) {
+      return false;
+    }
+    // Cannot remove an element unless version is higher for all other actors as well.
+    if (!dominates(version, model.values.get(value.getId()).version)) {
+      return false;
+    }
+    model.version.put(key, clockValue);
+    model.values.remove(value.getId());
+    return true;
+  }
+
+  private Map<String, VersionedValue<T>> mergeItems(CollectionData<T> data1, CollectionData<T> data2) {
+    Map<String, VersionedValue<T>> merged = new HashMap<>();
+    for (VersionedValue<T> v2 : data2.values.values()) {
+      if (model.values.containsKey(v2.value.getId())) {
+        merged.put(v2.value.getId(), new VersionedValue(
+            v2.value, mergeVersions(model.values.get(v2.value.getId()).version, v2.version)));
+      } else if (!dominates(data1.version, v2.version)) {
+        merged.put(v2.value.getId(), new VersionedValue(v2.value, v2.version));
+      }
+    }
+    for (VersionedValue<T> v1 : data1.values.values()) {
+      if (!data2.values.containsKey(v1.value.getId()) && !dominates(data2.version, v1.version)) {
+        merged.put(v1.value.getId(), new VersionedValue(v1.value, v1.version));
+      }
+    }
+    return merged;
+  }
+
+  private VersionMap mergeVersions(VersionMap version1, VersionMap version2) {
+    VersionMap merged = new VersionMap();
+    for (Map.Entry<String, Integer> entry : version1.entrySet()) {
+      merged.put(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, Integer> entry : version2.entrySet()) {
+      Integer version1Value = version1.get(entry.getKey());
+      merged.put(entry.getKey(), Math.max(entry.getValue().intValue(),
```

Comment:
I wish we could also port https://github.com/PolymerLabs/arcs/blob/master/src/runtime/crdt/tests/crdt-collection-tests.ts to gain confidence! Soon...

Reply:
yessss, coming soon :)